src/core/Form/SearchInput/SearchInput.tsx

```diff
@@ -1,47 +1,265 @@
+  onChange?: (value: StateValue) => void;
+  /** Callback for onBlur event */
+  onBlur?: (event: FocusEvent<HTMLInputElement>) => void;
+  /** Callback for search button click */
+  onSearch?: (value: StateValue) => void;
+}
 
 const baseClassName = 'fi-search-input';
-const iconBaseClassName = `${baseClassName}_icon`;
+const searchInputClassNames = {
+  fullWidth: `${baseClassName}--full-width`,
+  error: `${baseClassName}--error`,
+  notEmpty: `${baseClassName}--not-empty`,
+  inputElement: `${baseClassName}_input`,
+  inputElementContainer: `${baseClassName}_input-element-container`,
+  functionalityContainer: `${baseClassName}_functionality-container`,
+  button: `${baseClassName}_button`,
+  searchButton: `${baseClassName}_button-search`,
+  searchIcon: `${baseClassName}_button-search-icon`,
+  clearButton: `${baseClassName}_button-clear`,
+  clearIcon: `${baseClassName}_button-clear-icon`,
+};
 
-export interface SearchInputProps extends Omit<TextInputProps, 'status'> {
-  status?: 'default' | 'error';
+interface SearchInputState {
+  value: string | number | string[] | undefined;
+}
+
+class BaseSearchInput extends Component<SearchInputProps> {
+  state: SearchInputState = {
+    value: this.props.value || this.props.defaultValue || '',
+  };
+
+  private inputRef = createRef<HTMLInputElement>();
+
+  private id: string;
+
+  private statusTextId: string;
+
+  static getDerivedStateFromProps(
+    nextProps: SearchInputProps,
+    prevState: SearchInputState,
+  ) {
+    const { value } = nextProps;
+    if ('value' in nextProps && value !== prevState.value) {
+      return { value };
+    }
+    return null;
+  }
+
+  constructor(props: SearchInputProps) {
+    super(props);
+    this.id = `${idGenerator(props.id)}`;
+    this.statusTextId = `${this.id}-statusText`;
+  }
+
+  render() {
+    const {
+      value,
+      defaultValue,
+      className,
+      labelText,
+      labelMode,
+      clearButtonLabel,
+      searchButtonLabel,
+      searchButtonProps,
+      inputContainerProps,
+      onChange: propOnChange,
+      onSearch: propOnSearch,
+      children,
+      status,
+      statusText,
+      visualPlaceholder,
+      id,
+      fullWidth,
+      'aria-describedby': ariaDescribedBy,
+      ...passProps
+    } = this.props;
+
+    const conditionalSetState = (newValue: StateValue) => {
+      if (!('value' in this.props)) {
+        this.setState({ value: newValue });
+      }
+      if (propOnChange) {
+        propOnChange(newValue);
+      }
+    };
+
+    const onSearch = () => {
+      if (!!propOnSearch) {
+        propOnSearch(this.state.value);
+      }
+    };
+
+    const onClear = () => {
+      conditionalSetState('');
+      setTimeout(() => {
+        if (this.inputRef.current) {
+          this.inputRef.current.focus();
+        }
+      }, 100);
+    };
+
+    const onKeyPress = (event: React.KeyboardEvent<HTMLInputElement>) => {
+      if (!!this.state.value && event?.key === 'Enter') {
```

Comment:
Should we allow searching with empty strings?

Reply:
Actually, we can't since the design pattern does not allow it either.

Reply:
That's something I'd leave to whoever writes the search function anyway.

Reply:
Cannot be done. Design blocks it since we show the activated button and clear button only after content has been added. Or we have to activate the search button onFocus and disable it onBlur which does not leave the decision for the users/developers either.

Reply:
Good point. No searching with empty strings then.